src/librustdoc/clean/mod.rs

```diff
@@ -418,9 +418,11 @@ impl<'tcx> Clean<Type> for ty::ProjectionTy<'tcx> {
             GenericBound::TraitBound(t, _) => t.trait_,
             GenericBound::Outlives(_) => panic!("cleaning a trait got a lifetime"),
         };
+        let self_type = self.self_ty().clean(cx);
         Type::QPath {
             name: cx.tcx.associated_item(self.item_def_id).ident.name,
-            self_type: box self.self_ty().clean(cx),
+            self_def_id: self_type.def_id(),
+            self_type: box self_type,
```

Comment:
Why add a separate `def_id` field? You can just call `self_type.def_id()` wherever you need it.

Reply:
because `self_type` is a `Generic(Symbol("Self"))` most of the time, and it was much easier to pass is that way. I also tried to add a new `Type` called `SelfType`, but it was much more work to implement.

Reply:
I personally feel a bit uncomfortable with the added field, would it make sense to add a helper function/method somewhere instead?

Reply:
The only reason the field exists that the `DefId` of the `Self` type is lost after cleaning because the `Self` generic parameter is resolved to `Generic("Self")`, which does not contain any `DefId`.
I don't know how a new helper function would help with that.

Reply:
Ahh, sorry, I misread your above comment. Request withdrawn.